python/src/nanoarrow/array.py

```diff
@@ -0,0 +1,243 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from functools import cached_property
+from typing import Iterable
+
+from nanoarrow._lib import CDEVICE_CPU, CArray, CDevice, CMaterializedArrayStream
+from nanoarrow.c_lib import c_array, c_array_stream
+from nanoarrow.iterator import iterator
+from nanoarrow.schema import Schema
+
+
+class Scalar:
+    """Generic wrapper around an :class:`Array` element
```

Comment:
> Actually, I didn't think of it when I wrote this, but the `CScalar` probably should _not_ implement `__arrow_c_array__` because it is not necessarily representing a CPU array.

Yeah, it's true that things are not fully correct regarding devices right now, but I think that is something we can tackle later. i.e. when there is better device support, we would add `__arrow_c_device_array__` in addition, which always works, and let `__arrow_c_array__` raise an error or something like that.

The same happens in pyarrow actually. A pyarrow.Array can right now be backed by non-CPU data, but we still have a `_arrow_c_array__` defined on it.